deps/v8/src/sampler.cc

```diff
@@ -342,6 +342,14 @@ void SignalHandler::HandleProfilerSignal(int signal, siginfo_t* info,
     // We require a fully initialized and entered isolate.
     return;
   }
+
+  // ISSUE-14576: To avoid deadlock, return if there is a thread lock
+  if (Isolate::GetProcessWideMutexPointer()->TryLock()) {
```

Comment:
@dmelikyan According to different man pages (like http://man7.org/linux/man-pages/man2/sigaction.2.html, but also on OSX, I haven't looked on other platforms) and other sources, including "Advanced Programming In the Unix Environment" from Richard Stevens, it seems that `sigaction`:

> guarantee[s] that whenever we are processing a given signal, another occurrence of that same signal is blocked until we're finished processing the first occurrence. 

(that's an excerpt from the book mentioned above).

So I would think that your approach is safe.

However @tunniclm mentioned in [one of his comments](https://github.com/joyent/node/pull/25309#issuecomment-109214872) that he investigated:

> Block[ing] the SIGPROF signal using pthread_sigmask() while acquiring and releasing the process_wide_mutex_

which suggests that `SIGPROF` is not blocked when `SignalHandler::HandleProfilerSignal` runs. @tunniclm Are we missing something?